report.go

```diff
@@ -0,0 +1,162 @@
+package gomoss
+
+import (
+	"fmt"
+	"github.com/anaskhan96/soup"
+	"net/url"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// Region specifies similarity from line a to line b
+type Region struct {
+	from int64
+	to   int64
+}
+
+// Source is structure that containing plagiarism result of a source code
+type Source struct {
+	Filename   string
+	Similarity string
+	Regions    []Region
+}
+
+// Match means compare the plagiarism between two source codes
+type Match struct {
+	Left, Right Source
+}
+
+// Report is data extracted data from Moss URL
+type Report struct {
+	Matches []Match
+}
+
+// FindLinks find all analysis related hrefs in HTML
+func findLinks(ref *url.URL) []string {
+	resp, _ := soup.Get(ref.String())
+
+	doc := soup.HTMLParse(resp)
+	anchors := doc.FindAll("a")
+	exist := make(map[string]bool)
+	var links []string
+	for _, anchor := range anchors {
+		link := anchor.Attrs()["href"]
+		if strings.Contains(link, "match") && !exist[link] {
+			exist[link] = true
+			links = append(links, link)
+		}
+	}
+	return links
+}
+
+// DownloadURL downloads URL to file
+func DownloadURL(ref *url.URL, filename string) error {
+	resp, err := soup.Get(ref.String())
+	if err != nil {
+		return err
+	}
+
+	f, err := os.Create(filename)
+	defer f.Close()
+	if err != nil {
+		return err
+	}
+
+	reg := regexp.MustCompile(`http:\/\/moss.stanford.edu\/results\/\d*\/`)
+	resp = reg.ReplaceAllString(resp, "")
+
+	_, err = f.Write([]byte(resp))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Download downloads all HTML files from given URL recursively
+func Download(ref *url.URL) error {
+	basedir := "moss.stanford.edu"
+	err := os.MkdirAll(basedir, 0755)
+	if err != nil {
+		return fmt.Errorf("Can't create a directory")
+	}
+
+	links := findLinks(ref)
+	ref.Path = "index.html"
```

Comment:
instead of changing `ref.Path`, it might be better to do something like

```go
func mustParse(url *url.URL, rawurl string) *url.URL {
  u, err := url.Parse(rawurl)
  if err != nil {
    panic(err)
  }
  return u
}

...
DownloadURL(mustParse(ref, "index.html"), path.Join(basedir, "index.html"))
```